haskell/ghc_bindist_hadrian.bzl

```diff
@@ -0,0 +1,439 @@
+        all_outputs.append(out)
+        ctx.actions.run_shell(
+            outputs=[out],
+            inputs=depset([f]),
+            arguments=[f.path, out.path],
+            command="mkdir -p $(dirname $2) && cp $1 $2")
+
+    return [
+        DefaultInfo(
+            files=depset(all_outputs),
+            runfiles=ctx.runfiles(files=all_outputs))
+    ]
+
+
+copy_filegroups_to_this_package = rule(
+    implementation=_copy_filegroup_impl,
+    attrs={
+        "srcs": attr.label_list(),
+    },
+)
+
+# Targets are of the form x86_64-unknown-linux.
+def _split_target(target):
+    arch, _, os = target.split("-")
+    return (arch, os)
+
+# Main function.
+# It fills the template ghc.BUILD.tpl with the targets specific to the platform the build is done on.
+def _ghc_bindist_hadrian_impl(ctx):
+    filepaths = resolve_labels(ctx, [
+        "@rules_haskell//haskell:ghc.BUILD.tpl",
+        "@rules_haskell//haskell:private/pkgdb_to_bzl.py",
+    ])
+    _, os = _split_target(ctx.attr.target)
+    unpack_dir = ""
+
+    ctx.download_and_extract(
+        url = ctx.attr.url,
+        output = unpack_dir,
+        sha256 = ctx.attr.sha256,
+        type = "tar.xz",
+        stripPrefix = ctx.attr.strip_prefix,
+    )
+
+    bindir = "bin"
+    libdir = "lib"
+    docdir = "docs"
+
+    # The bindist requires patching invalid haddock paths. See Makefile of GHC:
+    # https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-9.2.3-release/hadrian/bindist/Makefile#L54-74
+    result = ctx.execute(["bash", "-c", """\
+set -euo pipefail
+find {lib}/package.conf.d -name "*.conf" -print0 | \\
+  xargs -0 sed -i.bak 's|\\${{pkgroot}}/\\.\\./\\.\\.|${{pkgroot}}/..|'
+find {lib}/package.conf.d -name "rts-*.conf" -print0 | \\
+  xargs -0 sed -i.bak2 's|haddock-html:.*$|haddock-html:|'
+""".format(
+        lib = libdir,
+    )])
+    if result.return_code != 0:
+        fail(result.stderr)
+
+    # Since Bazel regular rules generate files in the "execroots" and GHC requires some files to be next to eachothers,
+    # one has to move all the files coming from the GHC bindist tarball (which is unpacked by a repository rule) to an execroot.
+    # These copied versions of the files are the 'generated_*_filegroup' targets.
+
+
+    generated_bin_filegroup = define_rule(
+        "copy_filegroups_to_this_package",
+        name = "generated_bin_filegroup",
+        srcs = [":bin"],
+    )
+
+    generated_lib_filegroup = define_rule(
+        "copy_filegroups_to_this_package",
+        name = "generated_lib_filegroup",
+        srcs = [":lib"],
+    )
+
+    generated_docdir_filegroup = define_rule(
+        "copy_filegroups_to_this_package",
+        name = "generated_docdir_filegroup",
+        srcs = [":{}".format(docdir)],
+    )
+
+    generated_include_filegroup = define_rule(
+        "copy_filegroups_to_this_package",
+        name = "generated_include_filegroup",
+        srcs = [":{}".format(docdir)],
+    )
```

Comment:
You can move these into [`haskell/toolchain.bzl` next to `_hadrian_bindist_settings`](https://github.com/tweag/rules_haskell/pull/1842/files#diff-f83278238101cead941a012a809724f4fcac6997d0106d70678372f5c533ca84R590) as
```
copy_filegroups_to_this_package(
    name = "generated_bin_filegroup",
    srcs = [":bin"],
)
...
```
and the definition of `copy_filegroups_to_this_package` as well.

Reasons why that may be preferable:
1. The `ghc.BUILD.tpl` is also used by the other bindist toolchains and needs to remain valid for those use cases until we can supersede them with the Hadrian one. But, the `%{generated_*_filegroup}` placeholders are only used for the Hadrian case. Invoking the copy-files rules within `haskell_toolchain` macro means that these placeholders are no longer needed.
2. It's preferable to not mix repository rule and regular rule definitions in the same file, so that changes to the rule definition don't also invalidate the repository rule and cause unnecessary refetching. If the copy-file rule is moved into `haskell_toolchain` then the repository rules and regular rules are separated and the import of `ghc_bindist_hadrian.bzl` in `ghc.BUILD.tpl` is also no longer needed.
3. It puts both concerns into the same place: generating the settings file and placing the bindist files next to it.